packages/rollup-core/test/app/canonical-chain-batch-finalizer.spec.ts

```diff
@@ -0,0 +1,186 @@
+/* External Imports */
+import { keccak256FromUtf8, TestUtils, sleep } from '@eth-optimism/core-utils'
+import { JsonRpcProvider, TransactionReceipt } from 'ethers/providers'
+
+/* Internal Imports */
+import {
+  CanonicalChainBatchFinalizer,
+  DefaultDataService,
+} from '../../src/app/data'
+import { BatchSubmission, BatchSubmissionStatus } from '../../src/types/data'
+
+class MockDataService extends DefaultDataService {
+  public throwOnFinalize: boolean = false
+  public getWasInvoked: boolean = false
+  public batchSubmissions: BatchSubmission[] = []
+  public finalized: Map<number, BatchSubmission> = new Map<
+    number,
+    BatchSubmission
+  >()
+
+  constructor() {
+    super(undefined)
+  }
+
+  public async getNextCanonicalChainTransactionBatchToFinalize(): Promise<
+    BatchSubmission
+  > {
+    this.getWasInvoked = true
+    return this.batchSubmissions.shift()
+  }
+
+  public async markTransactionBatchFinalOnL1(
+    batchNumber: number,
+    submissionTxHash: string
+  ): Promise<void> {
+    if (this.throwOnFinalize) {
+      throw Error('You told me to...')
+    }
+
+    this.finalized.set(batchNumber, {
+      batchNumber,
+      submissionTxHash,
+      status: BatchSubmissionStatus.FINALIZED,
+    })
+  }
+}
+
+class MockProvider extends JsonRpcProvider {
+  public txReceipts: Map<string, TransactionReceipt> = new Map<
+    string,
+    TransactionReceipt
+  >()
+
+  public async waitForTransaction(
+    hash: string,
+    numConfirms: number
+  ): Promise<TransactionReceipt> {
+    while (!this.txReceipts.get(hash)) {
+      await sleep(100)
+    }
+    return this.txReceipts.get(hash)
+  }
+}
+
+describe('Canonical Chain Batch Finalizer', () => {
+  let batchFinalizer: CanonicalChainBatchFinalizer
+  let dataService: MockDataService
+  let provider: MockProvider
+  beforeEach(async () => {
+    dataService = new MockDataService()
+    provider = new MockProvider()
+    batchFinalizer = new CanonicalChainBatchFinalizer(
+      dataService,
+      provider,
+      1,
+      10
+    )
+  })
+
+  it('should run successfully when no batch is built', async () => {
+    const res = await batchFinalizer.runTask()
+    res.should.equal(false, 'No batch should have been finalized')
+    dataService.getWasInvoked.should.equal(
+      true,
+      'getNextCanonicalChainTransactionBatchToFinalize not invoked!'
+    )
+  })
+
+  it('should throw if returned batch submission is not SENT', async () => {
+    const submissionTxHash: string = keccak256FromUtf8('tx1')
+    const batchNumber: number = 0
+    dataService.batchSubmissions.push({
+      batchNumber,
+      submissionTxHash,
+      status: BatchSubmissionStatus.QUEUED,
+    })
+
+    provider.txReceipts.set(submissionTxHash, { status: 1 } as any)
+
+    TestUtils.assertThrowsAsync(async () => {
```

Comment:
Might be worth checking the error message matches the relevant implementation's msg:
```
Received tx batch to finalize in ${
        batchToFinalize.status
      } instead of ${
        BatchSubmissionStatus.SENT
      }. Batch Submission: ${JSON.stringify(batchToFinalize)}.
```

Reply:
I'll create a new `Error` that I can match on. Matching text is brittle and will break if the message is updated even if the logic works.